### PR TITLE
Scrub compaction: add a new in-memory partition segregation method

### PR DESCRIPTION
**Pull request by denesb**
The current disk-based segregation method works well enough for most cases, but it struggles greatly when there are a lot of partitions in the input. When this is the case it will produce tons of buckets (sstables), in the order of hundreds or even thousands. This puts a huge strain on different parts of the system.
This series introduces a new segregation method which specializes on the lots of small partitions case. If the conditions are right, it can cause a drastic reduction of buckets. In one case I tested, a 1.1GB sstable with 3.6M partitions in it produced just 2 output sstables, down from the 500+ with the on-disk method.
This new method uses a memtable to sort out-of-order partitions. In-order partitions bypass this sorting altogether and go to the disk directly. This method is not suitable for cases where either the partition are large or the total amount of data is large. For those the disk-based method should be used. Scrub compaction decides on the method to use based on heuristics.

Tests: unit(dev)

### CODE REVIEW

File: mutation_writer/partition_based_splitting_writer.cc [hunk 1]
```diff
@@ -20,11 +20,114 @@
...
+
+private:
+    void init_current_mut(dht::decorated_key key, tombstone tomb) {
+        _current_mut_size = key.external_memory_usage();
+        _current_mut.emplace(_schema, std::move(key));
+        _current_mut->partition().apply(tomb);
+    }
+
+    future<> flush_memtable() {
+        co_await _consumer(_memtable->make_flush_reader(_schema, _permit, _pc));
```
Comment (avikivity) [inline mutation_writer/partition_based_splitting_writer.cc:50]: A little disappointing that we need to duplicate this code (and can't just ask the memtable to flush itself), but it's not a big deal. The duplication isn't large and it's still playing with lego bricks.

File: mutation_writer/partition_based_splitting_writer.hh [hunk 2]
```diff
@@ -27,17 +27,40 @@
...
+// reached. In-order partitions bypass this and go to an output sstable
+// directly.
+// Use when the input has lots of small partition and a small or moderate amount
+// of data in total (in the order of megabytes).
+struct in_memory_segregate_config {
+    // For flushing the memtable which does the in-memory segregation (sorting)
+    // part.
+    const io_priority_class& pc;
+    // Maximum amount of memory to be used by the in-memory segregation
+    // (sorting) structures. Partitions can be split across partitions
```
Comment (avikivity) [inline mutation_writer/partition_based_splitting_writer.hh:41]: across sstables

File: mutation_writer/partition_based_splitting_writer.hh [hunk 3]
```diff
@@ -27,17 +27,40 @@
...
+// internally. When a new partition can't be written into any of the existing
+// buckets, open a new one (see \ref on_disk_segregate_config::max_buckets).
+struct on_disk_segregate_config {
+    // Maximum number of live buckets. When reaching the limit, an existing (the
+    // largest) bucket will be closed before a new one is created.
+    unsigned max_buckets;
+};
+
+using segregate_method_config = std::variant<in_memory_segregate_config, on_disk_segregate_config>;
+
```
Comment (avikivity) [inline mutation_writer/partition_based_splitting_writer.hh:56]: Do we need the two methods? memtables seem strictly superior.

File: compaction/compaction.cc [hunk 4]
```diff
@@ -1399,6 +1399,32 @@ class scrub_compaction final : public regular_compaction {
     std::string _scrub_finish_description;
     uint64_t _bucket_count = 0;
```
Comment (avikivity) [inline compaction/compaction.cc:1400]: I don't understand the conditions. Let's say the total amount of data is 1TB and partitions are 1GB. Why not use memtables? Provided we can close+open, which I think we can.

**Comment by avikivity** [comment]
Very nice and clean, but I didn't understand the rationale to use on-disk sorting.

File: mutation_writer/partition_based_splitting_writer.cc (hunk 1 above)
Comment (denesb) [reply mutation_writer/partition_based_splitting_writer.cc:50]: Actually here it is an advantage that flush happens via a reader. The consumer is abstracted, we know it is an sstable, but the code here doesn't.

**Comment by denesb** [comment]
> Very nice and clean, but I didn't understand the rationale to use on-disk sorting.

I'm going to answer this here, instead of splitting the discussion across the two related comments you made.

Let's define two terms to help the discussion:
* *Out-of-order set* (OOO set) the part of the data which is out-of-order. A partition is a member of the OOO set if it is out-of-order w.r.t. the previous partition in the data. Note that partitions in the OOO set may be in-order/out-of-order w.r.t. each other.
* *Entropy* - the amount of partitions in the OOO set which are out-of-order w.r.t. each other. An OOO set with an entropy factor of 0 means that all partitions in said set are in-order w.r.t. each other. The data stream can simply be split into two in-order data stream (main set and OOO set). An entropy factor of 1 effectively means that partitions in the OOO set are in reverse order.

I have two reasons for keeping the on-disk sorting: large amount of total data and large partitions:
* Large amount of total data. This will cause many output sstables to be created because we need to flush the memtable a lot due to the possibly large OOO set. However I'm not really sure the disk-based sorting would do better here, depends entirely on the entropy factor.
* Large partitions. For the purposes of this discussion we consider a partition large if it is larger than max_memory. When partitions cross this threshold, the memtable will cease sorting and instead it will split partitions further. On the other hand if partitions are large, there tends to be less of them.

Number of sstables produced for a given scenario/method/entropy factor:

| Large...   | Method   | Entropy factor 0                    | Entropy factor 1                         |
|------------|----------|-------------------------------------|------------------------------------------|
| Data       | Memtable | sizeof(OOO set)/max_memory + 1      | sizeof(OOO set)/max_memory + 1           |
| Data       | Disk     | 2                                   | cardinality(OOO set)                     |
| Partitions | Memtable | cardinality(OOO set)/max_memory + 1 | cardinality(OOO set)/max_memory + 1      |
| Partitions | Disk     | 2                                   | cardinality(OOO set)                     |

These are of course extremas, real life will be somewhere between 0-1 entropy and with variable data/partition size.
It is interesting that the amount of sstables the memtable method produces depends only on OOO set and max memory, and not on the entropy factor.

**Comment by avikivity** [comment]
> > Very nice and clean, but I didn't understand the rationale to use on-disk sorting.
> 
> I'm going to answer this here, instead of splitting the discussion across the two related comments you made.
> 
> Let's define two terms to help the discussion:
> 
> * _Out-of-order set_ (OOO set) the part of the data which is out-of-order. A partition is a member of the OOO set if it is out-of-order w.r.t. the previous partition in the data. Note that partitions in the OOO set may be in-order/out-of-order w.r.t. each other.
> * _Entropy_ - the amount of partitions in the OOO set which are out-of-order w.r.t. each other. An OOO set with an entropy factor of 0 means that all partitions in said set are in-order w.r.t. each other. The data stream can simply be split into two in-order data streams (main set and OOO set). An entropy factor of 1 effectively means that partitions in the OOO set are in reverse order.
> 
> I have two reasons for keeping the on-disk sorting: large amount of total data and large partitions:
> 
> * Large amount of total data. This will cause many output sstables to be created because we need to flush the memtable a lot due to the possibly large OOO set. However I'm not really sure the disk-based sorting would do better here, depends entirely on the entropy factor.

Since we don't know the entropy, and the user has no way to guess it, best to prepare for the worse case. What's the user supposed to do, try out all methods?


> * Large partitions. For the purposes of this discussion we consider a partition large if it is larger than max_memory. When partitions cross this threshold, the memtable will cease sorting and instead it will split partitions further. On the other hand if partitions are large, there tends to be less of them.

What's the problem with that? Yes, read amplification for that partition will increase until it is compacted. But it should be compacted soon (and since those sstables won't be large, it will be quite soon).


> 
> Number of sstables produced for a given scenario/method/entropy factor:
> 
> Large...	Method	Entropy factor 0	Entropy factor 1
> Data	Memtable	sizeof(OOO set)/max_memory + 1	sizeof(OOO set)/max_memory + 1
> Data	Disk	2	cardinality(OOO set)
> Partitions	Memtable	cardinality(OOO set)/max_memory + 1	cardinality(OOO set)/max_memory + 1
> Partitions	Disk	2	cardinality(OOO set)
> These are of course extremas, real life will be somewhere between 0-1 entropy and with variable data/partition size. It is interesting that the memtable method is independent of the entropy factor, while the disk method is independent of the size of the OOO set.


This code isn't going to be heavily used. In fact there's a chance that it will never be used. So it's best to keep things simple so we don't have to test multiple variants.

Note that since you have the special case for in-order partitions, I think memtables will work better than on-disk even for low entropy (and the statement that memtable is independent of entropy is incorrect).

**Comment by denesb** [comment]
> > > Very nice and clean, but I didn't understand the rationale to use on-disk sorting.
> > 
> > 
> > I'm going to answer this here, instead of splitting the discussion across the two related comments you made.
> > Let's define two terms to help the discussion:
> > 
> > * _Out-of-order set_ (OOO set) the part of the data which is out-of-order. A partition is a member of the OOO set if it is out-of-order w.r.t. the previous partition in the data. Note that partitions in the OOO set may be in-order/out-of-order w.r.t. each other.
> > * _Entropy_ - the amount of partitions in the OOO set which are out-of-order w.r.t. each other. An OOO set with an entropy factor of 0 means that all partitions in said set are in-order w.r.t. each other. The data stream can simply be split into two in-order data streams (main set and OOO set). An entropy factor of 1 effectively means that partitions in the OOO set are in reverse order.
> > 
> > I have two reasons for keeping the on-disk sorting: large amount of total data and large partitions:
> > 
> > * Large amount of total data. This will cause many output sstables to be created because we need to flush the memtable a lot due to the possibly large OOO set. However I'm not really sure the disk-based sorting would do better here, depends entirely on the entropy factor.
> 
> Since we don't know the entropy, and the user has no way to guess it, best to prepare for the worse case. What's the user supposed to do, try out all methods?

No, the choice is made by scrub based on the available data.

> 
> > * Large partitions. For the purposes of this discussion we consider a partition large if it is larger than max_memory. When partitions cross this threshold, the memtable will cease sorting and instead it will split partitions further. On the other hand if partitions are large, there tends to be less of them.
> 
> What's the problem with that? Yes, read amplification for that partition will increase until it is compacted. But it should be compacted soon (and since those sstables won't be large, it will be quite soon).

Problems currently only arise when too many sstables are created: in the order of hundreds or even thousands. This only starts to become a problem once there is a partition with size in the 100GB range.

> 
> > Number of sstables produced for a given scenario/method/entropy factor:
> > Large...	Method	Entropy factor 0	Entropy factor 1
> > Data	Memtable	sizeof(OOO set)/max_memory + 1	sizeof(OOO set)/max_memory + 1
> > Data	Disk	2	cardinality(OOO set)
> > Partitions	Memtable	cardinality(OOO set)/max_memory + 1	cardinality(OOO set)/max_memory + 1
> > Partitions	Disk	2	cardinality(OOO set)
> > These are of course extremas, real life will be somewhere between 0-1 entropy and with variable data/partition size. It is interesting that the memtable method is independent of the entropy factor, while the disk method is independent of the size of the OOO set.
> 
> This code isn't going to be heavily used. In fact there's a chance that it will never be used. So it's best to keep things simple so we don't have to test multiple variants.
> 
> Note that since you have the special case for in-order partitions, I think memtables will work better than on-disk even for low entropy (and the statement that memtable is independent of entropy is incorrect).

Ok, I will remove it. Note that entropy as I defined it only refers to the OOO set itself, and once you consider it like that the memtable is independent of it because it will not create more or less output sstables if the entropy is lower or higher.

**Comment by denesb** [comment]
v2:
* Drop disk-based segregation method

**Pull request closed without merging**